### ISSUE TITLE
unpassed optional parameter turned into null

### ISSUE THREAD

**Issue by Diluka** [typeorm/typeorm#3852]
**Issue type:**

[ ] question
[x] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

```ts
  async isLeader(activityId: number, playerId: number, isActivated?: boolean) {
    console.log(arguments) // [Arguments] { '0': 1, '1': 1 }
    console.log({ activityId, leaderId: playerId, isActivated }) // { activityId: 1, leaderId: 1, isActivated: undefined }
    const count = await this.count({ activityId, leaderId: playerId, isActivated });
    return count > 0;
  }
```

```sql
SELECT COUNT(DISTINCT(`Group`.`id`)) as "cnt" FROM `group` `Group` WHERE `Group`.`activityId` = ? AND `Group`.`leaderId` = ? AND `Group`.`isActivated` = ? -- PARAMETERS: [1,1,null]
```

## Expect
- when `isActivated` is passed `true` shall generate sql `isActivated = TRUE`
- when `isActivated` is passed `false` shall generate sql `isActivated = FALSE`
- when `isActivated` is passed `null` shall generate sql `isActivated IS NULL`
- when `isActivated` is not passed shall generate **nothing**

**Comment by hcesar** [comment]
There is no such thing as `not passed` in typescript.  The value is there `isActivated: undefined`. Is is equivalent of `x.isLeader(1,1, undefined)`.

If TypeORM is treating `undefined` as `null`, I bet it will not get changed, otherwise it would be a huge compatibility breaking change.

A workaround would be testing the value before sending to typeorm. 
Something like this would do the tricky:
```
const query = { activityId, leaderId: playerId, ...(isActivated !== undefined && {isActivated}) };
const count = await this.count(query);
```

Query would only have `isActivated` property if is not `undefined`

**Comment by Diluka** [comment]
@hcesar `undefined` is not equivalent to `null` in javascript.
`undefined` means not exist. `null` means empty.
`undefined === null` is `false`

btw, in sql `isActivated = NULL` is wrong

**Comment by den-t** [comment]
Agree that it would be a breaking change. Could it be put into 1.x then?

I second @Diluka that generating `column = NULL` for an undefined property doesn't make sense.

There must be a reason why it's done like that. But it eludes me.
There was a comment from @Kononnable in [#2500](https://github.com/typeorm/typeorm/issues/2500#issuecomment-454560399). And their position was to treat undefined differently to null.

Could the core team, please, share their thoughts?

E.g. currently:
```
const where = { name: "Sample" };
await repository.find({ where });
// WHERE "Entity"."name" = ? -- PARAMETERS: ["Sample"]

where.name = null;
await repository.find({ where });
// WHERE "Entity"."name" IS NULL

where.name = undefined;
await repository.find({ where });
// WHERE "Entity"."name" = ? -- PARAMETERS: [null]
```
Expected the last query to not have a where clase.

**Comment by hcesar** [comment]
> @hcesar `undefined` is not equivalent to `null` in javascript.
> `undefined` means not exist. `null` means empty.
> `undefined === null` is `false`
> 
> btw, in sql `isActivated = NULL` is wrong

Where have I said null is equivalent to undefined?
The point here is that the property do exist, and its value is undefined.

How TypeORM is dealing with that might be a problem, but it is there.

**Comment by den-t** [comment]
Yep, property exists. And is passed through, unlike the issue title suggests.
We might benefit from renaming the issue. 🤗
